src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE;
+    private static final int MIN_SIZE = 0;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+        size = MIN_SIZE;
+    }
+
+    private Object[] grow() {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity - MAX_CAPACITY > 0) {
+            newCapacity = MAX_CAPACITY;
+        }
+        return Arrays.copyOf(elementData, newCapacity);
+    }
+
+    private void checkRangeForAdd(int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Illegal index: " + index);
+        }
+    }
+
+    private void checkRange(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Illegal index: " + index);
+        }
+    }
+
+    private T fastRemote(int index) {
+        T value = (T) elementData[index];
+        if (size - 1 > index) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - 1);
+        }
+        size--;
+        return value;
+    }
+
+    private boolean arrayListEquals(Object first, Object second) {
+        return (first == second) || (first != null && first.equals(second));
+    }
```

Comment:
Let's move this method closer to its first usage. Also, could be renamed to `elementsEqual`.